src/dune/build_system.ml

```diff
@@ -1360,59 +1409,78 @@ end = struct
+        select_sandbox_mode ~loc:rule_loc
           (Dep.Set.sandbox_config deps)
           ~sandboxing_preference:t.sandboxing_preference
     in
-    let action_ctx = Action_exec.Context.make ~targets ~context ~env in
-    let rule_digest =
-      let env = Action_exec.Context.env action_ctx in
-      let trace =
-        ( Dep.Set.trace deps ~sandbox_mode ~env ~eval_pred
-        , List.map targets_as_list ~f:(fun p -> Path.to_string (Path.build p))
-        , Option.map context ~f:(fun c -> c.name)
-        , Action.for_shell action )
+    let* rule_need_rerun =
+      let force =
+        !Clflags.force
+        && List.exists targets_as_list ~f:Path.Build.is_alias_stamp_file
       in
-      Digest.generic trace
+      let always_rebuild = Dep.Set.has_universe deps in
+      if force || always_rebuild then
+        Fiber.return true
+      else
+        (* [prev_trace] will be [None] if rule is run for the first time. *)
+        let prev_trace = Trace_db.get (Path.build head_target) in
+        (* [targets_digest] will be [None] if not all targets were build. *)
+        let targets_digest = compute_targets_digest targets_as_list in
+        let rule_digest =
+          compute_rule_digest rule ~deps ~action ~sandbox_mode
+        in
+        let rule_or_targets_changed =
+          match (prev_trace, targets_digest) with
+          | Some prev_trace, Some targets_digest ->
+            prev_trace.rule_digest <> rule_digest
+            || prev_trace.targets_digest <> targets_digest
+          | _ -> true
+        in
+        if rule_or_targets_changed then
+          Fiber.return true
+        else
+          let prev_trace = Option.value_exn prev_trace in
+          let rec loop stages =
+            match stages with
+            | [] -> Fiber.return false
+            | (deps, old_digest) :: rest ->
+              let* () = build_deps deps in
+              let new_digest =
+                compute_dependencies_digest deps ~sandbox_mode ~env ~eval_pred
+              in
+              if old_digest <> new_digest then
+                Fiber.return false
+              else
+                loop rest
+          in
+          loop prev_trace.dynamic_deps_stages
     in
-    let targets_digest = compute_targets_digest targets_as_list in
     let sandbox =
       Option.map sandbox_mode ~f:(fun mode ->
-        let digest = Digest.to_string rule_digest in
-        (Path.Build.relative sandbox_dir digest, mode))
-    in
-    let force =
-      !Clflags.force
-      && List.exists targets_as_list ~f:Path.Build.is_alias_stamp_file
-    in
-    let something_changed =
-      Dep.Set.has_universe deps
-      ||
-      match (prev_trace, targets_digest) with
-      | Some prev_trace, Some targets_digest ->
-        prev_trace.rule_digest <> rule_digest
-        || prev_trace.targets_digest <> targets_digest
-      | _ -> true
+        let sandbox_suffix =
+          compute_rule_digest rule ~deps ~action ~sandbox_mode
```

Comment:
Are we computing this value twice? (isn't rule_digest below the same) We probably shouldn't.

Reply:
It was computed not twice but even three times. Fixed.